core/src/main/java/org/apache/accumulo/core/data/ImmutableByteSequence.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.accumulo.core.data;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.Objects;
+
+import org.apache.accumulo.core.util.ByteBufferUtil;
+
+/**
+ * An immutable implementation of {@link ByteSequence} that copies data in its constructor and never
+ * lets its internal byte array escape. This class is final to prevent it from being made mutable in
+ * a subclass.
+ *
+ * <p>
+ * This class is package private because direct construction is not expected, instead use of methods
+ * like {@link ByteSequence#of(byte[])} is how instances of this should be created.
+ * </p>
+ */
+final class ImmutableByteSequence extends ByteSequence {
+
+  private static final long serialVersionUID = 1L;
+
+  // A reference to this must never escape this class.
+  private final byte[] data;
+
+  private int hash = 1;
+  private boolean hashIsOne = false;
+
+  ImmutableByteSequence(ByteSequence data) {
+    if (data.isBackedByArray()) {
+      this.data =
+          Arrays.copyOfRange(data.getBackingArray(), data.offset(), data.offset() + data.length());
+    } else {
+      var dataArray = data.toArray();
+      this.data = Arrays.copyOf(dataArray, dataArray.length);
```

Comment:
There is a private static method , `ArrayByteSequence.copy`, that has very simliar logic. I wonder if that method should be made public and  moved into `ByteSequence`. Then this becomes:
```
  this.data = ByteSequence.copy(data);
```

Reply:
The else of this method and the other method have a difference in that one cares if the byte array ref escapes and the other does not.  Made some changes in b52c685 related to this to avoid two allocations.

As for adding a copy method to public API there are already two way in public API to copy. For immutable copy the new `of` method.  For mutable copy the existing ArrayByteSequence constructor.